apps/src/labs/ProjectManager.ts

```diff
@@ -0,0 +1,145 @@
+import {SourcesStore} from './SourcesStore';
+import {ChannelsStore} from './ChannelsStore';
+import {Project} from './types';
+
+export enum ProjectManagerEvent {
+  SaveStart,
+  SaveNoop,
+  SaveSuccess,
+  SaveFail
+}
+
+export default class ProjectManager {
+  channelId: string;
+  sourcesStore: SourcesStore;
+  channelsStore: ChannelsStore;
+  getProject: () => Project;
+
+  private nextSaveTime: number | null = null;
+  private readonly saveInterval: number = 30 * 1000; // 30 seconds
+  private saveInProgress = false;
+  private saveQueued = false;
+  private eventListeners: {
+    [key in keyof typeof ProjectManagerEvent]?: [(response: Response) => void];
+  } = {};
+
+  constructor(
+    channelId: string,
+    sourcesStore: SourcesStore,
+    channelsStore: ChannelsStore,
+    getProject: () => Project
+  ) {
+    this.channelId = channelId;
+    this.sourcesStore = sourcesStore;
+    this.channelsStore = channelsStore;
+    this.getProject = getProject;
+  }
+
+  async load(): Promise<Response> {
```

Comment:
Have you thought about how you want to handle DB-backed projects (no channel ID)? Will they also go through the project manager or will they use a separate system?

Reply:
Good question. Assuming we continue to support DB-backed projects (I think we will?) I think it makes sense to extend this model to include those types of projects. Which may require some reworking of the interface.

Reply:
I looked into this more today. There's a key differences I see with level sources vs project levels: projects are unrelated to progress, and milestones are posted separately from project saves. However, level sources saving is a part of milestone reporting, and only happens when a [milestone is reported](https://github.com/code-dot-org/code-dot-org/blob/54787e2ee9d8c2e930e9f618b8d97b065c82f5c2/dashboard/app/controllers/activities_controller.rb#L68) (vs. autosave for project levels). This makes me inclined to keep the `ProjectManager` strictly for code saved as part of a 'project', i.e. unrelated to a milestone. We could rename `channelId` here to be something more generic so it more clearly applies to the local storage version.